Attempting to assign to a function name produces uncompilable C++

### Current Behavior:

The following Stan model compiles in stanc3:

```stan
functions {
   real foo(int x){
     return 1.0;
   }

   real bar(int y){
     return 2.0;
   }

   void baz(){
     // NB: putting this in later blocks produce an error due to assigning to a global variable
     foo = bar;
   }
}

model {
     print(foo(1));
    baz();
   print(foo(1));
}
```

but the C++ fails to compile:

<details><summary>Cmdstan output</summary>

```shell
Building fun.stan
make: Entering directory '/home/brian/Dev/cpp/cmdstan'
cd /home/brian/Dev/ml/stanc3 && echo "--- Rebuilding stanc ---\n" && dune build @install
--- Rebuilding stanc ---\n
cp /home/brian/Dev/ml/stanc3/_build/default/src/stanc/stanc.exe bin/stanc

--- Translating Stan model to C++ code ---
bin/stanc  --o=../../ml/stanc3/fun.hpp ../../ml/stanc3/fun.stan

--- Compiling, linking C++ code ---
g++ -std=c++1y -pthread -D_REENTRANT -Wno-sign-compare -Wno-ignored-attributes      -I stan/lib/stan_math/lib/tbb_2020.3/include    -O3 -I src -I stan/src -I lib/rapidjson_1.1.0/ -I lib/CLI11-1.9.1/ -I stan/lib/stan_math/ -I stan/lib/stan_math/lib/eigen_3.3.9 -I stan/lib/stan_math/lib/boost_1.75.0 -I stan/lib/stan_math/lib/sundials_6.0.0/include -I stan/lib/stan_math/lib/sundials_6.0.0/src/sundials    -DBOOST_DISABLE_ASSERTS          -c -Wno-ignored-attributes   -x c++ -o ../../ml/stanc3/fun.o ../../ml/stanc3/fun.hpp
In file included from stan/lib/stan_math/stan/math/prim/meta/is_eigen_matrix_base.hpp:6,
                 from stan/lib/stan_math/stan/math/prim/meta/is_eigen.hpp:5,
                 from stan/lib/stan_math/stan/math/prim/meta/base_type.hpp:6,
                 from stan/lib/stan_math/stan/math/prim/meta/return_type.hpp:6,
                 from stan/lib/stan_math/stan/math/prim/meta/append_return_type.hpp:5,
                 from stan/lib/stan_math/stan/math/prim/meta.hpp:176,
                 from stan/lib/stan_math/stan/math/rev/core/accumulate_adjoints.hpp:4,
                 from stan/lib/stan_math/stan/math/rev/core.hpp:4,
                 from stan/lib/stan_math/stan/math/rev.hpp:8,
                 from stan/lib/stan_math/stan/math.hpp:19,
                 from stan/src/stan/model/model_header.hpp:4:
stan/lib/stan_math/stan/math/prim/meta/is_base_pointer_convertible.hpp: In instantiation of ‘struct stan::is_base_pointer_convertible<Eigen::EigenBase, double (&)(const int&, std::basic_ostream<char>*)>’:
stan/lib/stan_math/stan/math/prim/meta/is_var_matrix.hpp:18:8:   recursively required by substitution of ‘template<class T> struct stan::value_type<T, typename std::enable_if<stan::is_eigen<T>::value, void>::type> [with T = double (&)(const int&, std::basic_ostream<char>*)]’
stan/lib/stan_math/stan/math/prim/meta/is_var_matrix.hpp:18:8:   required from ‘struct stan::is_var_matrix<double (&)(const int&, std::basic_ostream<char>*)>’
stan/lib/stan_math/stan/math/prim/meta/conjunction.hpp:17:8:   required from ‘struct stan::math::conjunction<stan::is_var_matrix<double (&)(const int&, std::basic_ostream<char, std::char_traits<char> >*)>, stan::is_eigen<double (&)(const int&, std::basic_ostream<char, std::char_traits<char> >*)> >’
stan/lib/stan_math/stan/math/prim/meta/require_helpers.hpp:63:24:   required by substitution of ‘template<class ... Checks> using require_any_not_t = std::enable_if_t<(! stan::math::conjunction<T>::value)> [with Checks = {stan::is_var_matrix<double (&)(const int&, std::basic_ostream<char, std::char_traits<char> >*)>, stan::is_eigen<double (&)(const int&, std::basic_ostream<char, std::char_traits<char> >*)>}]’
stan/src/stan/model/indexing/access_helpers.hpp:43:65:   required by substitution of ‘template<class T1, class T2, stan::require_any_not_t<stan::is_var_matrix<Container>, stan::is_eigen<T_actual> >* <anonymous> > void stan::model::internal::assign_impl(T1&&, T2&&) [with T1 = double (&)(const int&, std::basic_ostream<char>*); T2 = double (&)(const int&, std::basic_ostream<char>*); stan::require_any_not_t<stan::is_var_matrix<Container>, stan::is_eigen<T_actual> >* <anonymous> = <missing>]’
stan/src/stan/model/indexing/assign.hpp:57:24:   required from ‘void stan::model::assign(T&&, U&&, const char*) [with T = double (&)(const int&, std::basic_ostream<char>*); U = double (&)(const int&, std::basic_ostream<char>*); stan::require_t<std::is_assignable<typename std::decay<_Tp>::type&, typename std::decay<_Func>::type> >* <anonymous> = 0]’
../../ml/stanc3/fun.hpp:73:61:   required from here
stan/lib/stan_math/stan/math/prim/meta/is_base_pointer_convertible.hpp:29:17: error: no matching function for call to ‘stan::is_base_pointer_convertible<Eigen::EigenBase, double (&)(const int&, std::basic_ostream<char>*)>::f(double (*)(const int&, std::basic_ostream<char>*))’
   29 |     = decltype(f(std::declval<std::remove_reference_t<Derived> *>()))::value
      |                ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
stan/lib/stan_math/stan/math/prim/meta/is_base_pointer_convertible.hpp:24:26: note: candidate: ‘static std::false_type stan::is_base_pointer_convertible<Base, Derived>::f(const void*) [with Base = Eigen::EigenBase; Derived = double (&)(const int&, std::basic_ostream<char>*); std::false_type = std::integral_constant<bool, false>]’ <near match>
   24 |   static std::false_type f(const void *);
      |                          ^
stan/lib/stan_math/stan/math/prim/meta/is_base_pointer_convertible.hpp:24:26: note:   conversion of argument 1 would be ill-formed:
stan/lib/stan_math/stan/math/prim/meta/is_base_pointer_convertible.hpp:29:17: error: invalid conversion from ‘double (*)(const int&, std::basic_ostream<char>*)’ to ‘const void*’ [-fpermissive]
   29 |     = decltype(f(std::declval<std::remove_reference_t<Derived> *>()))::value
      |                ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                 |
      |                 double (*)(const int&, std::basic_ostream<char>*)
stan/lib/stan_math/stan/math/prim/meta/is_base_pointer_convertible.hpp:26:25: note: candidate: ‘template<class OtherDerived> static std::true_type stan::is_base_pointer_convertible<Base, Derived>::f(const Base<OtherDerived>*) [with OtherDerived = OtherDerived; Base = Eigen::EigenBase; Derived = double (&)(const int&, std::basic_ostream<char>*)]’
   26 |   static std::true_type f(const Base<OtherDerived> *);
      |                         ^
stan/lib/stan_math/stan/math/prim/meta/is_base_pointer_convertible.hpp:26:25: note:   template argument deduction/substitution failed:
stan/lib/stan_math/stan/math/prim/meta/is_base_pointer_convertible.hpp:29:17: note:   mismatched types ‘const Eigen::EigenBase<Derived>’ and ‘double(const int&, std::basic_ostream<char>*)’
   29 |     = decltype(f(std::declval<std::remove_reference_t<Derived> *>()))::value
      |                ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make: *** [make/program:58: ../../ml/stanc3/fun] Error 1
make: Leaving directory '/home/brian/Dev/cpp/cmdstan'
```
</details>


### Expected Behavior:

At the moment I would expect this to be an error. After something like #742 or lambdas, I would expect the C++ generated to compile.


This is probably a low-priority issue as the cases it comes up are obviously pathological, I just wanted to document it.

--- a/src/frontend/Semantic_error.ml
+++ b/src/frontend/Semantic_error.ml
@@ -345,6 +345,7 @@ module StatementError = struct
   type t =
     | CannotAssignToReadOnly of string
     | CannotAssignToGlobal of string
+    | CannotAssignFunction of UnsizedType.t * string
     | LValueMultiIndexing
     | InvalidSamplingPDForPMF
     | InvalidSamplingCDForCCDF of string
@@ -380,6 +381,9 @@ module StatementError = struct
         Fmt.pf ppf
           "Cannot assign to global variable '%s' declared in previous blocks."
           name
+    | CannotAssignFunction (ut, name) ->
+        Fmt.pf ppf "Cannot assign a function type '%a' to variable '%s'."
+          UnsizedType.pp ut name
     | LValueMultiIndexing ->
         Fmt.pf ppf
           "Left hand side of an assignment cannot have nested multi-indexing."
@@ -657,6 +661,9 @@ let cannot_assign_to_read_only loc name =
 
 let cannot_assign_to_global loc name =
   StatementError (loc, StatementError.CannotAssignToGlobal name)
+
+let cannot_assign_function loc ut name =
+  StatementError (loc, StatementError.CannotAssignFunction (ut, name))
 
 let cannot_assign_to_multiindex loc =
   StatementError (loc, StatementError.LValueMultiIndexing)

--- a/src/frontend/Semantic_error.mli
+++ b/src/frontend/Semantic_error.mli
@@ -119,6 +119,7 @@ val empty_array : Location_span.t -> t
 val bad_int_literal : Location_span.t -> t
 val cannot_assign_to_read_only : Location_span.t -> string -> t
 val cannot_assign_to_global : Location_span.t -> string -> t
+val cannot_assign_function : Location_span.t -> UnsizedType.t -> string -> t
 val cannot_assign_to_multiindex : Location_span.t -> t
 val invalid_sampling_pdf_or_pmf : Location_span.t -> t
 val invalid_sampling_cdf_or_ccdf : Location_span.t -> string -> t

--- a/src/frontend/Typechecker.ml
+++ b/src/frontend/Typechecker.ml
@@ -937,6 +937,15 @@ let verify_assignment_global loc cf block is_global id =
   if (not is_global) || block = cf.current_block then ()
   else Semantic_error.cannot_assign_to_global loc id.name |> error
 
+(* Until function types are added to the user language, we
+   disallow assignments to function values
+*)
+let verify_assignment_non_function loc ut id =
+  match ut with
+  | UnsizedType.UFun _ | UMathLibraryFunction ->
+      Semantic_error.cannot_assign_function loc ut id.name |> error
+  | _ -> ()
+
 let check_assignment_operator loc assop lhs rhs =
   let err op =
     Semantic_error.illtyped_assignment loc op lhs.lmeta.type_ rhs.emeta.type_
@@ -1019,6 +1028,7 @@ let check_assignment loc cf tenv assign_lhs assign_op assign_rhs =
         |> error in
   verify_assignment_global loc cf block global assign_id ;
   verify_assignment_read_only loc readonly assign_id ;
+  verify_assignment_non_function loc rhs.emeta.type_ assign_id ;
   let rhs' = check_assignment_operator loc assign_op lhs rhs in
   mk_typed_statement ~return_type:NoReturnType ~loc
     ~stmt:(Assignment {assign_lhs= lhs; assign_op; assign_rhs= rhs'})

--- a/test/integration/bad/assigment-to-function-stanmath.stan
+++ b/test/integration/bad/assigment-to-function-stanmath.stan
@@ -1,0 +1,16 @@
+functions {
+   real foo(real x, real y){
+     return 1.0;
+   }
+
+   void baz(){
+     // NB: putting this in later blocks produce an error due to assigning to a global variable
+     foo = pow;
+   }
+}
+
+model {
+     print(foo(2,3));
+    baz();
+   print(foo(2,3));
+}

--- a/test/integration/bad/assignment-to-function.stan
+++ b/test/integration/bad/assignment-to-function.stan
@@ -1,0 +1,20 @@
+functions {
+   real foo(int x){
+     return 1.0;
+   }
+
+   real bar(int y){
+     return 2.0;
+   }
+
+   void baz(){
+     // NB: putting this in later blocks produce an error due to assigning to a global variable
+     foo = bar;
+   }
+}
+
+model {
+     print(foo(1));
+    baz();
+   print(foo(1));
+}

--- a/test/integration/bad/stanc.expected
+++ b/test/integration/bad/stanc.expected
@@ -79,6 +79,18 @@ Semantic error in 'array_expr_bad2.stan', line 2, column 40 to column 41:
    -------------------------------------------------
 
 Array expression must have entries of consistent type. Expected array[] int but found int.
+  $ ../../../../install/default/bin/stanc assigment-to-function-stanmath.stan
+Semantic error in 'assigment-to-function-stanmath.stan', line 8, column 5 to column 15:
+   -------------------------------------------------
+     6:     void baz(){
+     7:       // NB: putting this in later blocks produce an error due to assigning to a global variable
+     8:       foo = pow;
+              ^
+     9:     }
+    10:  }
+   -------------------------------------------------
+
+Cannot assign a function type '<Stan Math function>' to variable 'foo'.
   $ ../../../../install/default/bin/stanc assign_real_to_int.stan
 Semantic error in 'assign_real_to_int.stan', line 11, column 2 to column 8:
    -------------------------------------------------
@@ -90,6 +102,18 @@ Semantic error in 'assign_real_to_int.stan', line 11, column 2 to column 8:
    -------------------------------------------------
 
 Ill-typed arguments supplied to assignment operator =: lhs has type int and rhs has type real
+  $ ../../../../install/default/bin/stanc assignment-to-function.stan
+Semantic error in 'assignment-to-function.stan', line 12, column 5 to column 15:
+   -------------------------------------------------
+    10:     void baz(){
+    11:       // NB: putting this in later blocks produce an error due to assigning to a global variable
+    12:       foo = bar;
+              ^
+    13:     }
+    14:  }
+   -------------------------------------------------
+
+Cannot assign a function type '(int) => real' to variable 'foo'.
   $ ../../../../install/default/bin/stanc bad_fundef_returntypes1.stan
 Semantic error in 'bad_fundef_returntypes1.stan', line 2, column 13 to line 6, column 3:
    -------------------------------------------------

